framework/src/base/Moose.C

```diff
@@ -27,6 +27,7 @@
 const ExecFlagType EXEC_NONE = registerDefaultExecFlag("NONE");
 const ExecFlagType EXEC_INITIAL = registerDefaultExecFlag("INITIAL");
 const ExecFlagType EXEC_LINEAR = registerDefaultExecFlag("LINEAR");
+const ExecFlagType EXEC_NONLINEAR_CONVERGENCE = registerDefaultExecFlag("NONLINEAR_CONVERGENCE");
```

Comment:
We really at least need a documentation on what this flag is, or when this flag is executed because users will see this. Number of execute flags will blow up if we keep adding. This design is upside down imho. This requirement can be achieved by using `Executioner::addAttributeReporter` and let the executioner assign the value to the reporter programmatically. If we can retrieve the residual norm from SNES, we can also save one residual evaluation.

Reply:
doco should go on this page @joshuahansel 
https://mooseframework.inl.gov/source/interfaces/SetupInterface.html

Are you talking about setting the value of the residual from the executioner?
This is not enough unfortunately. we want to be able to define convergence based on postprocessors that are not just this "Residual" object. To do this, they must be executed on this special execution flag otherwise they are either executed too often (LINEAR) or too late (NONLINEAR)

Reply:
Yes, thank you, I forgot about documenting the flag. I'll add some doc and make it as visible as I can.

For your design concerns, does Guillaume's explanation answer it? This PR is a pre-req for the upcoming Convergence system, where users can dial up their own convergence criteria. One such Convergence object will be one that compares a PP value to a tolerance. The `Residual` PP changes I made here were kind of just a test, but also could conceivably be used in some convergence criteria in the future. The main motivator for this flag is to be able to execute objects such as PPs on nonlinear iterations, but not `NONLINEAR`, which is too late, since it's after it has been decided that there will be another nonlinear step. The possible cascading dependencies of an object like a PP means that a lot of systems could be involved and need this flag, which is why I went with making it a default flag.

Reply:
See https://github.com/idaholab/moose/pull/28806 for an example of a PP that can be used in convergence criteria.

Reply:
Ok, I've added the new flag to `SetupInterface.md`. I also changed the `execute_on` description method to drop the list of options (since they're redundant in doc now) and provide a URL to that `SetupInterface.md` page.

Reply:
This can be resolved. Because this flag is never used by some custom executioners, like the Richardson iteration executioner in Griffin, having it seen by users is not ideal. To me, execute flags should be tight with executioners, except initial, residual, jacobian, final, etc., which is the reason I suggested moving the definition of this flag in FEProblemSolve, but my opinion on that is not as strong because we are not following that rule anyway.

Reply:
>Are you talking about setting the value of the residual from the executioner?

Yes, we had a way to expose some important info of executioner into reporters directly.

Reply:
I see your point about certain flags not being relevant in certain executioners. When we have an object that may be used by used in any executioner, I don't see how you can get around registering the union of all those flags for a given application. I think the best you could do is throw an error in your executioner if a user supplies an irrelevant flag. But I agree, it's kind of a mess, but I don't have an alternative design idea.

Reply:
We could add the ability to dereguster exec flags with automated substituions or an error